lib/srv/desktop/rdp/rdpclient/src/ga.rs

```diff
@@ -0,0 +1,164 @@
+//!
+//! This constraint is enforced because the underlying memory is expected to
+//! be passed back into Rust from Go via FFI compatible functions, which might
+//! be called from any arbitrary thread selected by Go's runtime. By enforcing
+//! [`Send`] + [`Sync`], we can guarantee that the underlying memory is always
+//! safe to send to and access from any thread.
+//!
+//! # Example
+//!
+//! ```no_run
+//! use ga::{GoArc_new, GoArc_clone, GoArc_drop};
+//!
+//! /// Imagine [`create_go_arc`] being called by CGO via the FFI.
+//! /// Go holds on to the [`GoArc`] returned to it by that call,
+//! /// uses it in any arbitrary goroutine(s) via [`use_go_arc`],
+//! /// and then drops it when it's done with it via [`drop_go_arc`].
+//!
+//! /// Creates a new [`GoArc`] containing the value 42.
+//! #[no_mangle]
+//! pub unsafe extern "C" fn create_go_arc() -> GoArc<u32> {
+//!     GoArc_new(42)
+//! }
+//!
+//! /// Uses the given [`GoArc`] by cloning it into a [`Arc`].
+//! ///
+//! /// This can be called from any arbitrary goroutine(s) in Go,
+//! /// so long as [`drop_go_arc`] hasn't been called yet.
+//! #[no_mangle]
+//! pub unsafe extern "C" fn use_go_arc(go_arc: GoArc<u32>) -> GoArc<u32> {
+//!     // Clone the [`GoArc`] into a [`Arc`] so that we can
+//!     // safely access the value inside.
+//!     let rust_arc = GoArc_clone(go_arc);
+//!
+//!     // We can use the [`Arc`] just like a normal [`Arc`] here.
+//!     // For example, we can directly call any non-mutating methods on
+//!     // the underlying u32 via the [`Deref`] trait.
+//!     let is_lt = rust_arc.lt(32);
+//!     if is_lt {
+//!        println!("{} is less than {}", 32, rust_arc);
+//!     } else {
+//!         println!("{} is not less than {}", 32, rust_arc);
+//!     }
+//!
+//!     // We can even clone it again if we want to, pass it around
+//!     // to other threads, etc.
+//!     let rust_arc_clone = rust_arc.clone();
+//!     std::thread::spawn(move || {
+//!        println!("Hello from another thread! {}", rust_arc_clone);
+//!     });
+//! }
+//!
+//! /// Drops the given [`GoArc`].
+//! ///
+//! /// This must be called once and only once by Go when it's done
+//! /// with the [`GoArc`]. After this is called, [`use_go_arc`] can
+//! /// no longer be called.
+//! #[no_mangle]
+//! pub unsafe extern "C" fn drop_go_arc(go_arc: GoArc<u32>) {
+//!     // Drop the [`GoArc`] explicitly.
+//!     GoArc_drop(go_arc);
+//! }
+//!
+//!
+//! let go_arc = GoArc_new(42);
+//! ```
+use std::sync::Arc;
+
+/// A raw pointer to an [`Arc`] that is owned by Go.
+/// See the module level documentation for more details.
+pub type GoArc<T> = *mut Arc<T>;
+
+/// Creates a new [`GoArc`] containing the given value.
+#[allow(non_snake_case)]
+pub fn GoArc_new<T>(inner: T) -> GoArc<T>
+where
+    T: Send + Sync,
+{
+    // Create a new Send + Sync Arc
+    let arc = Arc::new(inner);
+    // Move it to the heap
+    let boxed_arc = Box::new(arc);
```

Comment:
What's the extra boxing of the pointer for? Can't we just use [`Arc::into_raw`](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.into_raw)/[`Arc::from_raw`](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.from_raw) and deal with a `*mut T` on the FFI side instead?

Reply:
As a bonus, that would let you use `&T` across the FFI boundary, which would prevent safe rust from calling your functions with some random garbage in the pointer.